File: algorithm_202012/baek4013.cpp
// baek4013 ATM

#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <algorithm>
#include <vector>
#include <utility>
#include <deque>

using namespace std;

typedef pair<int,int> pii;
typedef long long ll;

const int INF = 21 * 1e8;
const int MAX = 1e9;

int n,m;

vector<int> money;
vector<bool> rest;
vector<vector<int>> edge;
vector<int> discovered;
vector<vector<int>> scc;
vector<int> sccidx;
vector<bool> finished;
stack<int> st;
vector<int> earn;
vector<int> dp;

int s,P;
int discovcnt = 0;
int scccnt = 0;
int ans = 0;

int findscc(int cur)
{
   discovered[cur] = discovcnt++;
   int ret= discovered[cur];
   st.push(cur);

   for(auto next : edge[cur])
   {
       if(discovered[next] == -1)
       {
           ret= min(ret,findscc(next));
       }
       else if(!finished[next])
       {
           ret = min(ret,discovered[next]);
       }
   }    

   if(ret >= discovered[cur] && !finished[cur])
   {
       scc.push_back(vector<int>());
       while(1)
       {
           int stop = st.top();
           st.pop();

           money[scccnt] += earn[stop];
           sccidx[stop] = scccnt;
           scc[scccnt].push_back(stop);
           finished[stop] = true;

           if(stop == cur) break;
       }

       scccnt++;
   }

   return ret;
}

int main()
{
   ios_base :: sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
   cin>>n>>m;

   edge = vector<vector<int>>(n);
   rest = vector<bool>(n,false);
   discovered = vector<int>(n,-1);
   sccidx = vector<int>(n);
   finished = vector<bool>(n,false);
   earn = vector<int>(n,0);
   money = vector<int>(n,0);
   dp = vector<int>(n,0);
  
   for(int i = 0;i<m;i++)
   {
       int v1,v2;
       cin>>v1>>v2;
       v1--; v2--;
       edge[v1].push_back(v2);
   }

   for(int i = 0;i<n;i++)
   {
       cin>>earn[i];
   }

   cin>>s>>P;

   s--;

   for(int i = 0;i<P;i++)
   {
       int r;
       cin>>r;
       r--;

       rest[r] = true;
   }

   findscc(s);
   dp[scccnt-1] = money[scccnt-1];

   for(int i = scccnt-1;i>=0;i--)
   {
       for(auto elem : scc[i])
       {
           for(auto next : edge[elem])
           {
               if(sccidx[next] != i)
               {
                   dp[sccidx[next]] = max(dp[sccidx[next]],dp[i] + money[sccidx[next]]);
               }
           }
       }        
   }

   for(int i = 0;i<n;i++)
   {
       if(rest[i] && discovered[i] != -1)
       {
           ans = max(ans,dp[sccidx[i]]);
       }
   }

   cout<<ans;  

   return 0;
}

File: algorithm_202012/baek4013.py
# baek4013 ATM

import sys
from collections import deque
input=sys.stdin.readline
sys.setrecursionlimit(10**9)

n,m=map(int,input().split())
edges=[[] for _ in range(n+1)]
visited=[0 for _ in range(n+1)]
finished=[False for _ in range(n+1)]
SCCnum=[0 for _ in range(n+1)]
money=[0 for _ in range(n+1)]

for _ in range(m):
    a,b=map(int,input().split())
    edges[a].append(b)
for _ in range(1,n+1):
    money[_]=int(input())

start,rescnt=map(int,input().split())
restlist=list(map(int,input().split()))
moneyscc=[0 for _ in range(n+1)]

stack=[]
id = 0
SCCtotal=0
SCC=[]
def dfs(node ):
    global id,SCCtotal
    id += 1
    visited[node]= id  # node 별 dfs 순서 정의를 위해  0이면 아직 미 방문
    stack.append(node) # 스택에 삽입 - 같은 강한결합요소 체크를 위해
    parent =visited[node]
    for next in edges[node]:
        # 아직 방문하지 않았다면, dfs 수행 
        if visited[next] ==0: parent=min(parent, dfs(next))
        # 현재 dfs가 수행중인--> 부모일수도 아니면 형제일수도 
        elif not finished[next]: parent =min(parent,visited[next])
    
    # 부모 노드가 자기 자신인 경우
    if parent==visited[node] and not finished[node]:
        scc=[]
        while True:
            t=stack.pop()
            SCCnum[t]= SCCtotal
            moneyscc[SCCtotal]+=money[t]
            scc.append(t)
            finished[t]=True
            if t==node: break
        SCCtotal+=1
        SCC.append(scc)

    # 부모 출력
    return parent

dfs(start)

dp=[0 for _ in range(n+1)]

dp[SCCtotal-1]=moneyscc[SCCtotal-1]

for i in range(SCCtotal-1,-1,-1):
    for elem in SCC[i]:
        for next in edges[elem]:
            if SCCnum[next] != i:
                dp[SCCnum[next]] = max(dp[SCCnum[next]], dp[i]+ moneyscc[SCCnum[next]])
ans=0
for i in restlist:
    if visited[i]:
        ans=max(ans,dp[SCCnum[i]])   
print(ans)
